src/moonlink/src/row/column_array_builder.rs

```diff
@@ -95,9 +97,90 @@ impl ColumnArrayBuilder {
                 array_builder,
             ),
             DataType::List(inner) => ColumnArrayBuilder::new(inner.data_type(), capacity, true),
+            DataType::Struct(fields) => ColumnArrayBuilder::Struct(
+                StructBuilder::from_fields(fields.clone(), capacity),
+                array_builder,
+            ),
             _ => panic!("data type: {data_type:?}"),
         }
     }
+
+    fn append_value_to_struct_field(
+        builder: &mut StructBuilder,
+        i: usize,
+        value: &RowValue,
+    ) -> Result<(), Error> {
+        if let Some(bool_builder) = builder.field_builder::<BooleanBuilder>(i) {
+            match value {
+                RowValue::Bool(v) => bool_builder.append_value(*v),
+                RowValue::Null => bool_builder.append_null(),
+                _ => unreachable!("Bool expected from well-typed input"),
+            }
+        } else if let Some(int32_builder) = builder.field_builder::<PrimitiveBuilder<Int32Type>>(i)
+        {
+            match value {
+                RowValue::Int32(v) => int32_builder.append_value(*v),
+                RowValue::Null => int32_builder.append_null(),
+                _ => unreachable!("Int32 expected from well-typed input"),
+            }
+        } else if let Some(int64_builder) = builder.field_builder::<PrimitiveBuilder<Int64Type>>(i)
+        {
+            match value {
+                RowValue::Int64(v) => int64_builder.append_value(*v),
+                RowValue::Null => int64_builder.append_null(),
+                _ => unreachable!("Int64 expected from well-typed input"),
+            }
+        } else if let Some(float32_builder) =
+            builder.field_builder::<PrimitiveBuilder<Float32Type>>(i)
+        {
+            match value {
+                RowValue::Float32(v) => float32_builder.append_value(*v),
+                RowValue::Null => float32_builder.append_null(),
+                _ => unreachable!("Float32 expected from well-typed input"),
+            }
+        } else if let Some(float64_builder) =
+            builder.field_builder::<PrimitiveBuilder<Float64Type>>(i)
+        {
+            match value {
+                RowValue::Float64(v) => float64_builder.append_value(*v),
+                RowValue::Null => float64_builder.append_null(),
+                _ => unreachable!("Float64 expected from well-typed input"),
```

Comment:
nit: I would suggest to include value in the logging message, so if anything bad really happens we're still able to debug. example: https://github.com/Mooncake-Labs/moonlink/pull/1256